Font Library: add option to revoke access to Google Fonts

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Add an option to revoke access to Google Fonts to the the Font Library modal.

Partial for https://github.com/WordPress/gutenberg/issues/54826
(Partial because it only supports Google Fonts not other third parties).

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
There was no option for revoking access to Google Fonts once access was granted.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Adds a dropdown menu with an option that sets the 'wp-font-library-google-fonts-permission' local storage to false.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
Activate a block theme.
Open the styles sidebar in the Site Editor.
Open the Typography panel, then open the Font Library
Select the "Install Fonts" tab.

If you have not already granted permission to access Google Fonts, there should be no visible change when the PR is applied.

After allowing access to Google Fonts, there should be a button with an ellipsis to the right of the "Google Fonts" title.
The ellipsis should not show when you view a single font. Only on the first page and on the paged results.
Clicking the option should open a popup with a single option "Revoke access to Google Fonts".
Selecting this option should display the "Connect to Google Fonts" banner and button again.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<img width="976" alt="Font Library modal with the new ellipsis menu with the Revoke access option." src="https://github.com/WordPress/gutenberg/assets/7422055/db601eb5-33d8-4723-acf8-6e9233d8e4c4">

carolinan committed:
Font Library: add option to revoke access to Google Fonts

matiasbenedetto commented on packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js:
```diff
@@ -76,6 +77,14 @@ function FontCollection( { slug } ) {
 
+	const revokeAccess = () => {
+		window.localStorage.setItem(
+			'wp-font-library-google-fonts-permission',
```
We are using this string one more time. Could it be worth making it a constant?

matiasbenedetto commented on packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js:
```diff
@@ -223,11 +232,31 @@ function FontCollection( { slug } ) {
 		);
 	}
 
+	let actionsComponent = null;
+	if ( ! renderConfirmDialog && ! selectedFont ) {
+		actionsComponent = (
+			<DropdownMenu
+				icon={ moreVertical }
+				label={ __( 'Actions' ) }
+				popoverProps={ {
+					position: 'bottom left',
+				} }
+				controls={ [
+					{
+						title: __( 'Revoke access to Google Fonts' ),
+						onClick: revokeAccess,
+					},
+				] }
+			/>
+		);
+	}
+
```
Could this use a simple syntax? For example:

```jsx
const actionsComponent = () => {
  if (renderConfirmDialog || selectedFont) {
    return null;
  }
  return (
    <DropdownMenu
      icon={moreVertical}
      label={__("Actions")}
      popoverProps={{
        position: "bottom left",
      }}
      controls={[
        {
          title: __("Revoke access to Google Fonts"),
          onClick: revokeAccess,
        },
      ]}
    />
  );
};

```

matiasbenedetto commented on packages/edit-site/src/components/global-styles/font-library-modal/font-collection.js:
```diff
@@ -223,11 +232,31 @@ function FontCollection( { slug } ) {
 	}
 
+	let actionsComponent = null;
+	if ( ! renderConfirmDialog && ! selectedFont ) {
```
This should include checking whether the font collection rendered is the Google Fonts one. Otherwise, the component also renders on unwanted font collections (not the Google fonts one).

For example 
```js
slug === "google-fonts"
```

How it currently looks:
![image](https://github.com/WordPress/gutenberg/assets/1310626/70f117b8-e7b0-418c-9304-2f8913447640)

matiasbenedetto reviewed (commented):
Thanks @carolinan for working on this.
This is looking good, and it's partially working as expected. I added some comments in the code.

carolinan committed:
Change the ActionsComponent to a function component

Change the ActionsComponent to a function component.
Add a constant for the name of the Google Fonts local storage item.

carolinan committed:
Merge branch 'trunk' into update/font-library-revoke-access

carolinan committed:
Merge branch 'trunk' into update/font-library-revoke-access

carolinan commented:
@matiasbenedetto Thank you for the review. I believe I have addressed everything. I hope the logic is correct. The tests are also finally passing, so this is ready for another review.

matiasbenedetto committed:
Merge branch 'trunk' into update/font-library-revoke-access

matiasbenedetto reviewed (approved):
Thanks for this enhancement.
This is working as advertised.
LGTM

carolinan commented:
_If_ we consider this a requirement for the MVP, I think it needs to have the backport to beta/ RC label.

matiasbenedetto commented:
I found that this change is generating a visual regression: 

https://github.com/WordPress/gutenberg/assets/1310626/dd78f1fe-22ce-4c77-b08c-6be58f5f2440

carolinan commented:
I don't know how to test with the other third party collection. :(

matiasbenedetto commented:
> I don't know how to test with the other third party collection. :(

You can try just by adding a collection. Add this PHP somewhere. I.e. to theme  `functions.php` file:
```php


function register_my_collection() {
    $ubuntu = array(
        'font_family_settings' => array (
            'name'       => 'Ubuntu',
            'fontFamily' => 'Ubuntu, sans-serif',
            'slug'       => 'ubuntu',
        ),
        'categories' => array(
            'sans-serif',
        ),
    );
    
    $verdana = array(
        'font_family_settings' => array (
            'name'       => 'Verdana',
            'fontFamily' => 'Verdana, sans-serif',
            'slug'       => 'verdana',
        ),
        'categories' => array(
            'sans-serif',
        ),
    );

    $times = array(
        'font_family_settings' => array (
            'name'       => 'Times New Roman',
            'fontFamily' => 'Times New Roman, serif',
            'slug'       => 'times-new-roman',
        ),
        'categories' => array(
            'serif',
        ),
    );
    
    $categories = array(
        array (
            'name' => 'Sans Serif',
            'slug' => 'sans-serif',
        ),
        array (
            'name' => 'Serif',
            'slug' => 'serif',
        ),
    );

    $font_families = array ( $ubuntu, $verdana, $times );
    
    $collection = array(
        'name'          => 'My Custom Collection',
        'description'   => __( 'Custom fonts collection' ),
        'font_families' => $font_families,
        'categories'    => $categories
    );
    
    wp_register_font_collection( 'my-collection', $collection );
}

add_action( 'rest_api_init', 'register_my_collection' );

```

getdave commented:
I just _manually_ cherry-picked this PR to the [cherry-pick-wp-6-5-beta-3 branch](https://github.com/WordPress/gutenberg/pull/59394) to get it included in the next release.
